apollo-configservice/src/main/java/com/ctrip/framework/apollo/configservice/filter/ClientAuthenticationFilter.java

```diff
@@ -68,26 +70,49 @@ public void doFilter(ServletRequest req, ServletResponse resp, FilterChain chain
       return;
     }
 
+    // check timestamp and signature
     List<String> availableSecrets = accessKeyUtil.findAvailableSecret(appId);
     if (!CollectionUtils.isEmpty(availableSecrets)) {
       String timestamp = request.getHeader(Signature.HTTP_HEADER_TIMESTAMP);
       String authorization = request.getHeader(HttpHeaders.AUTHORIZATION);
+      String ip = WebUtils.tryToGetClientIp(request);
 
       // check timestamp, valid within 1 minute
       if (!checkTimestamp(timestamp)) {
-        logger.warn("Invalid timestamp. appId={},timestamp={}", appId, timestamp);
+        logger.warn("Invalid timestamp. appId={},clientIp={},timestamp={}", appId, ip, timestamp);
         response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "RequestTimeTooSkewed");
         return;
       }
 
       // check signature
-      String uri = request.getRequestURI();
-      String query = request.getQueryString();
-      if (!checkAuthorization(authorization, availableSecrets, timestamp, uri, query)) {
-        logger.warn("Invalid authorization. appId={},authorization={}", appId, authorization);
+      if (!checkAuthorization(authorization, availableSecrets, timestamp,
+          request.getRequestURI(), request.getQueryString())) {
+        logger.warn("Invalid authorization. appId={},clientIp={},authorization={}",
+            appId, ip, authorization);
         response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Unauthorized");
         return;
       }
+    } else {
+      // pre-check timestamp and signature
+      List<String> observableSecrets = accessKeyUtil.findObservableSecrets(appId);
```

Comment:
The observable check logic closely resembles the filter logic. Consider extracting it into a method that takes a mode variable.